tuplex/core/src/ee/local/LocalBackend.cc

```diff
@@ -2080,4 +2116,123 @@ namespace tuplex {
+            } else {
+                // put full partition to output set
+                limitedPartitions.push_back(partition);
+            }
+        }
+
+        // check the bottom output limit, adjust partitions if necessary
+        size_t numBottomOutputRows = 0;
+        size_t clippedBottom = 0;
+        for (auto it = partitions.rbegin(); it != partitions.rend(); it++) {
+            auto partition = *it;
+            numBottomOutputRows += partition->getNumRows();
+
+            if (partition == lastTopPart) {
+                // the bottom and the top partitions are overlapping
+                clippedBottom = bottomLimit - (numBottomOutputRows - partition->getNumRows());
+                if (clippedTop + clippedBottom >= partition->getNumRows()) {
+                    // if top and bottom range intersect, use full partitions
+                    clippedTop = partition->getNumRows();
+                    clippedBottom = 0;
+                }
+                break;
+            } else if (numBottomOutputRows >= bottomLimit) {
+                // clip last partition & leave loop
+                auto clipped = bottomLimit - (numTopOutputRows - partition->getNumRows());
+                assert(clipped <= partition->getNumRows());
+                Partition newPart = newPartitionWithSkipRows(partition, partition->getNumRows() - clipped, tstage);
+                partition->invalidate();
+                parition = newPart;
+                assert(partition->getNumRows() == clipped);
+                if (clipped > 0)
+                    limitedTailPartitions.push_back(partition);
+                break;
+            } else {
+                // put full partition to output set
+                limitedTailPartitions.push_back(partition);
+            }
+        }
+
+        // push the middle partition
+        if (lastTopPart != nullptr && (clippedTop > 0 || clippedBottom > 0)) {
+            assert(clippedTop + clippedBottom <= lastTopPart->getNumRows());
+
+            // split into two partitions with both top and bottom are in the same partition
+            Partition* lastBottomPart = nullptr;
+
+            if (clippedBottom != 0) {
+                lastBottomPart = newPartitionWithSkipRows(lastTopPart, lastTopPart->getNumRows() - clippedBottom, tstage);
+            }
+
+            lastTopPart->setNumRows(clippedTop);
+
+            limitedPartitions.push_back(lastTopPart);
+
+            if (lastBottomPart != nullptr) {
+                limitedPartitions.push_back(lastBottomPart);
+            }
+        }
+
+        // merge the head and tail partitions
+        partitions.clear()
+        partitions.insert(partitions.end(), limitedPartitions.begin(), limitedPartitions.end());
+        partitions.insert(partitions.end(), limitedTailPartitions.rbegin(), limitedTailPartitions.rend());
+    }
+
+    Partition* LocalBackend::newPartitionWithSkipRows(Partition *p_in, int numToSkip, TransformStage* tstage) {
```

Comment:
Maybe add Executor* as variable so it can be specified where to alloc partitions. (then just pass driver when you invoke this)

Reply:
TO create a test stage, use i.e. something like

```
    ContextOptions co = ContextOptions::defaults();
    auto enable_nvo = false; // test later with true! --> important for everything to work properly together!
    co.set("tuplex.optimizer.nullValueOptimization", enable_nvo ? "true" : "false");
    co.set("tuplex.useInterpreterOnly", "true");
    codegen::StageBuilder builder(0, true, true, false, 0.9, true, enable_nvo, false);
    auto csvop = FileInputOperator::fromCsv(test_path.toString(), co,
                                            option<bool>(true),
                                            option<char>(','), option<char>('"'),
                                            {""}, {}, {}, {});
    auto mapop = new MapOperator(csvop, UDF("lambda x: {'origin_airport_id': x['ORIGIN_AIRPORT_ID'], 'dest_airport_id':x['DEST_AIRPORT_ID']}"), csvop->columns());
    auto fop = new FileOutputOperator(mapop, test_output_path, UDF(""), "csv", FileFormat::OUTFMT_CSV, defaultCSVOutputOptions());
    builder.addFileInput(csvop);
    builder.addOperator(mapop);
    builder.addFileOutput(fop);

    auto tstage = builder.build();
```